### ISSUE TITLE
full or relative path?

### ISSUE THREAD

**Issue by gingerlime** [danmayer/coverband#188]
I'm finally looking at our first coverband reports on staging, and one thing I noticed is that when versions are deployed, coverband seems to treat files differently, even without changes to those files.

For example our app is deployed to `/var/local/company/company/` and we use symlinks to allow rollback etc, so the actual path is something like `/var/local/company/company.d/123` and the symlink points to it (i.e. `/var/local/company/company -> /var/local/company/company.d/123`)

When looking at the web-based coverage report, I can see some models that appear both as `app/models/my_model.rb` and `/var/local/company/company.d/123/app/models/my_model.rb` ...

Is there a way to consolidate those? or to only look at the relative path?

**Comment by kbaum** [comment]
Coverband has support configuring array of root_paths and it removes those root_paths with a `gsub` here:

https://github.com/danmayer/coverband/blob/master/lib/coverband/reporters/base.rb#L64

Could you try passing a regexp of your absolute root_path into the config?  Something like:
``` ruby
Coverband.configure do |config|
  config.root = [Dir.pwd, "/var/local/company/company.d/[0-9]*"]
end
```

**Comment by gingerlime** [comment]
Cool. Thanks @kbaum ! Just what I was looking for.

Should be `config.root_paths`, and also `Dir.pwd` should be [last](https://github.com/danmayer/coverband/blob/3130012b7daf7cce2c8324d4b79fe78d66958620/lib/coverband/reporters/base.rb#L72) ?
i.e.

```ruby
Coverband.configure do |config|
  config.root_paths = ["/var/local/company/company.d/[0-9]*", Dir.pwd]
end
```

But somehow, I'm still seeing two instances for each ruby file, and not entirely sure why or how to resolve this

![screen shot 2019-02-08 at 13 06 28](https://user-images.githubusercontent.com/524274/52461623-6e7ea280-2ba2-11e9-8c92-00cbb013b779.png)

I can't easily debug this on my dev box unfortunately, because paths are different there, so I'm a bit stuck on this one... It's seemingly a small thing, but throws off all our coverage stats plus lots of duplication unfortunately.

Any tips on how to troubleshoot this?

**Comment by gingerlime** [comment]
Update: this seems to work (trailing slash for the regex path)

```ruby
Coverband.configure do |config|
  config.root_paths = ["/var/local/company/company.d/[0-9]*/"]
end
```

But now there's a different problem, although I'm sure it's related, but somehow on every deploy, the stats look totally wonky

To reproduce:

* I deploy, clear coverage report, and make some requests so data is updated
* at this point coverage data looks good, makes sense

![screen shot 2019-02-08 at 14 19 31](https://user-images.githubusercontent.com/524274/52464301-d89c4500-2bac-11e9-950d-93bb77ad7a34.png)

* then I deploy again (without clearing the coverage report), and suddenly the data is all wrong / missing / inaccurate ... notice the relevant lines in particular, and lots of 100% that don't make sense, definitely weren't 100% covered...

![screen shot 2019-02-08 at 14 20 21](https://user-images.githubusercontent.com/524274/52464314-e782f780-2bac-11e9-8253-1502659f0964.png)

**Comment by kbaum** [comment]
Interesting.. One thing to note that if a file is simply required after coverband starts, then the loading of that file will cover much of the code. If it's a class for example, all the code hit when loading the ruby class will be hit including method definitions, class definitions, and constants. The method bodies will be missed. 

I also noticed on the demo that if all of the code is commented out in the file, there will be 0 relevant lines hit and 100% coverage.

https://coverband-demo.herokuapp.com/coverage?#478afe6df20062bdee1ac4f5c1edc44902a92b27

Could that help explain what's happening here?

**Comment by gingerlime** [comment]
hmmm... I'm not sure to be honest. If you look at the first screenshot, after clearing the report, it shows 8090 relevant lines, but then after deploy, only 190 relevant lines... So it looks like lots of code "went missing" (not sure how to describe it). It's simply missing from the coverage report.

I can see lots of our controllers, etc. Not sure what exactly is missing. Some controllers shouldn't get 100% coverage in my opinion (for example, rescue code that I'm pretty sure didn't execute), but yet they show as 100%

We're using puma with [prune_bundler](https://github.com/puma/puma/blob/36964ec42982d7b3205760bc2bf9ccf3fec8af69/examples/config.rb#L154-L159) for zero-downtime deploys. I'm guessing it's somehow related to this, but I would have expected coverband to also reload with the rest of the gems... ??

**Comment by kbaum** [comment]
Ok, i think i may know what's going on. If i open up my console and do:

```ruby
 puts Coverband.configuration.store.coverage.keys
```

I see absolute paths:

```
/Users/karlbaum/workspace/coverband/lib/coverband/adapters/base.rb
/Users/karlbaum/workspace/coverband/lib/coverband/adapters/redis_store.rb
/Users/karlbaum/workspace/coverband/lib/coverband/collectors/coverage.rb
/Users/karlbaum/workspace/coverband/lib/coverband/integrations/background.rb
/Users/karlbaum/workspace/coverband/lib/coverband.rb
/Users/karlbaum/workspace/coverband/lib/coverband/version.rb
/Users/karlbaum/workspace/coverband/lib/coverband/configuration.rb
/Users/karlbaum/workspace/coverband/lib/coverband/adapters/file_store.rb
/Users/karlbaum/workspace/coverband/lib/coverband/utils/s3_report.rb
```

Since your directory structure is changing each time you deploy, is your coverage being wiped? Can you execute this command and see what comes back?

```ruby
 puts Coverband.configuration.store.coverage.keys
```

**Comment by danmayer** [comment]
yeah feels like this has got to be related to the path normalization code... the regex on root paths was designed for that, but I haven't seen a system setup like that for awhile... I am guessing there is a bug... I can help look into it.

* some example keys as @kbaum mentioned would be helpful.
* your deployed paths if you could give two full path examples that match the regex
* full path for example file that is acting weird like a controller

I don't think you should have to do a coverband:clear ever as the new md5 of file changes takes care of that... if you do a clear AFTER a deploy it will make files hit during load disappear until next run. So that might explain a small part of this as well.... but things disappearing between deploys I belive is realted to the regex path support.

**Comment by danmayer** [comment]
if you share some more details I should have some time to look this weekend @gingerlime

**Comment by gingerlime** [comment]
@kbaum @danmayer Thank you!

Yes, I can see several folders listed, e.g.

```
/var/local/company/company.d/67/app/...
/var/local/company/company.d/68/app/...
```

Here's an output filtering for a specific controller (sanitized)

```
pp Coverband.configuration.store.coverage.keys.select { |k| k.match(/dashboard_controller/) }
["/var/local/company/company.d/78/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/74/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/72/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/71/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/70/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/69/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/68/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/67/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/66/app/controllers/dashboard_controller.rb"]
```

This controller now shows as having 100% coverage, but I'm quite confident that actual coverage is lower. There are some rescue clauses that aren't likely to get executed normally. I think the same goes to lots of other files.

Is this helpful? 

Side note:

> I haven't seen a system setup like that for awhile

I thought this was pretty standard for rails deployments with rollback support? but it's been a while since we've set this up. I'd love to switch to docker at some point, but things like zero-downtime deploys seem a bit tricky with docker (unless you go with Kubernetes, and feels like an overkill). In any case, always happy to learn something new, so that's why I'm asking :)

**Comment by kbaum** [comment]
> I thought this was pretty standard for rails deployments with rollback support?

This sounds like a solid approach to me. 99% of my rails deployment experience has been on Heroku which comes with zero downtime deploys using preboot.

**Comment by gingerlime** [comment]
I'm ashamed to admit, but I've never used Heroku. Of course I know lots of people use it, but I never did ... Both Unicorn and Puma support zero-downtime deploys (in various flavours, it seems), and also I thought that the `folder.d` structure with incremental releases and using symlink to point to the active and previous versions are pretty standard?

First time I hear of preboot as well :) It sounds like Heroku's preboot would launch a completely new VPS or containers for you, and would route to them using some kind of a load balancer, to do those zero-downtime deploys, but then this is probably different from creating essentially a cloned folder of rails with the new code, and then asking unicorn/puma to reload it.

In any case, I think our setup is definitely different from Heroku's, but not sure if it's vastly different from other non-heroku deployments out there (at least those that use Puma).

Sorry, not sure if this is contributing much to figuring out the problem here, but just not sure if we're doing something very different and unique, or fairly standard. I can't really say.

**Comment by danmayer** [comment]
Yeah I have seen the folder setup you have before, but between heroku and moving to docker for deployments. I used something very similar to your deploy process at my previous gig about 4 years ago.

I currently deploy with Docker and it works similarly for 0 downtime deployments as Heroku (basically a blue/green deployment process) where you fully boot up a new instance verify it's health checks and then move the load balancer to it, etc... We don't use Kubernetes but build on AWS ECS / Cloudformation so it still handles load balancer stuff for us.

Anyways, not saying the folder method is bad, I just haven't seen it or used it for years so I am guessing that a bug has been introduced related to modified deploy paths. I think all the data you provided should let me figure it out. 

In fact, I am pretty sure one bug is in the method below and how it plays with the roots regex config that is the issue... I am linking to my current branch as this method is already changing from the current release.

https://github.com/danmayer/coverband/blob/feature/gem_tracking/lib/coverband/reporters/base.rb#L70

You have a root like so `config.root_paths = ["/var/local/company/company.d/[0-9]*/"]` That code is trying to normalize all the varying filenames from previous deploys to match the currently deployed file structure... I am fairly certain it fails for your setup. I should be able to get a unit test in place with your data and update it to work.

**Comment by danmayer** [comment]
hmmm so @kbaum and @gingerlime this is a bit trickier, my first thought / tests I wasn't able to reproduce. We actually had a pretty good test for the regex file path stuff in place. I took data from this ticket but wasn't able to cause the error in my new test.

I was able to clean up the tests and improve the documentation a bit. Particularly around when one needs to use double quotes around the regex sent to `config.root_paths` (if using \d for example)

Here is the test changes I tried to reproduce with... take a look @gingerlime and let me know if anything looks different than your setup... If you are using a slightly different regex for example.

https://github.com/danmayer/coverband/commit/c10da72e1521e5ccebdbfa4020ab02e4a209642d

**Comment by danmayer** [comment]
Hmm so the prune_bundler and rolling restart could be interesting...

From the docs:
```
# is preserved across a phased-restart. (incompatible with preload_app)
# (off by default)
```
I am curious does this means in your production config you also don't eager load the app?
`config.eager_load = true` is that false perhaps?

I am thinking somehow on deploy... one of the following happens

* Coverage doesn't get started again
* Coverband doesn't get started again
* if not preloading that the baseline is screwed up...
* a bug in relevant lines https://github.com/danmayer/coverband/blob/master/lib/coverband/utils/source_file.rb#L200

For the files that show after the deploy... Are the normal line counts correct but all of the relevant line counts go to Zero or close to zero? IE does the controller in your screenshot only have 18 lines total? 

Might need to add some debug lines or get a partial dump of the coverage data for a single file... I am thinking we might actually handle the filenames correctly and merge the data, but if all the lines are marked not_relavent then it ignores the coverage data. @kbaum / @gingerlime outside of github or with scrubbing could you share a partial dataset of `Coverband.configuration.store.coverage['/full/path/to/dashboard_controller.rb'] possibly for 2 different deployed paths such as the last and the current deploy paths data? You should be able to email me and it doesn't include any source code info just an array of coverage hits.

**Comment by gingerlime** [comment]
> I am curious does this means in your production config you also don't eager load the app?
`config.eager_load = true` is that false perhaps?

It's set to `true` for production and staging, and `false` for development, test

> For the files that show after the deploy... Are the normal line counts correct but all of the relevant line counts go to Zero or close to zero? IE does the controller in your screenshot only have 18 lines total?

Yes, the controller on the screenshot does have 18 lines. Also other controllers show correct line counts as far as I can tell, including the `dashboard_controller` that I mentioned (that has some rescue clauses that I don't expect to be covered).

Here's the coverage data for this particular controller. I picked the earlier, latest and one in the middle

```
irb(main):002:0> Coverband.configuration.store.coverage['/***/66/app/controllers/dashboard_controller.rb']
=> {"first_updated_at"=>1549610119, "last_updated_at"=>1549610200, "file_hash"=>"14dc84e940e26cbfb9ac79b43862e762", "data"=>[1, 1, 1, nil, 1, 1, nil, nil, 1, nil, 1, 26, 26, nil, 26, 26, 26, 26, 26, 26, 26, nil, nil, 1, nil, 1, 26, 19, 0, 0, 0, 0, nil, nil, nil, nil, 0, 0, nil, nil, 1, 26, 26, 26, nil, nil, nil, nil, nil, 1, 26, nil, nil, 1, 26, nil, nil]}
irb(main):003:0> Coverband.configuration.store.coverage['/***/78/app/controllers/dashboard_controller.rb']
=> {"first_updated_at"=>1549658574, "last_updated_at"=>1549729830, "file_hash"=>"14dc84e940e26cbfb9ac79b43862e762", "data"=>[21, 21, 21, nil, 21, 21, nil, nil, 21, nil, 21, 22, 22, nil, 22, 22, 22, 22, 22, 22, 22, nil, nil, 21, nil, 21, 22, 13, 0, 0, 0, 0, nil, nil, nil, nil, 0, 0, nil, nil, 21, 22, 22, 22, nil, nil, nil, nil, nil, 21, 22, nil, nil, 21, 22, nil, nil]}
irb(main):004:0> Coverband.configuration.store.coverage['/***/70/app/controllers/dashboard_controller.rb']
=> {"first_updated_at"=>1549617873, "last_updated_at"=>1549618094, "file_hash"=>"14dc84e940e26cbfb9ac79b43862e762", "data"=>[16, 16, 16, nil, 16, 16, nil, nil, 16, nil, 16, 32, 32, nil, 32, 32, 32, 32, 32, 32, 32, nil, nil, 16, nil, 16, 32, 23, 0, 0, 0, 0, nil, nil, nil, nil, 0, 0, nil, nil, 16, 32, 32, 32, nil, nil, nil, nil, nil, 16, 32, nil, nil, 16, 32, nil, nil]}
```

Perhaps a stupid or silly question, because I know nothing about coverage or the internals of coverband: If you already keep an md5 hash of the file, couldn't you just use this hash to remove duplicates? So the path can be whatever the first (or latest) path might be, but if the file hash is the same, then the coverage report would show whichever path but only one entry per md5 signature? (Or perhaps you also want to remove the older version of xyz_controller.rb once it was updated? so you need to rely on the filename as well?)

side note:

> We don't use Kubernetes but build on AWS ECS / Cloudformation so it still handles load balancer stuff for us.

Are you able to share some more info about your setup? I've seen some guides, but there are always details that aren't being mentioned that I wonder about, e.g. handling db migrations, rollbacks, secrets, the entire docker build pipeline, sidekiq, gluing things together, etc. Those small bits that take it from a generic example to something that works in practice, with rails... Sorry it's a complete tangent, but I've been planning to move to Docker for production for ages, but get choice paralysis with all the options out there. Any help would be hugely appreciated.

**Comment by danmayer** [comment]
Thanks for sharing the data this should be helpful... It doesn't look like the coverage data would consider many of the lines not be "relevant" given that means Coverage should report them as nil and we see counts and 0... Makes me think something is happening when it tries to read and load the files...

Making regex explicit vs string or regex likely would be a bit safer. Would need to break backwards compatibility to do the change, thought and I haven't heard much usage of the regex directory feature for a few years. So hesitant to make to many changes there. I guess if we find the final bug is related to that I will be more inclined, but for now I think something else is causing the issue. 

> couldn't you just use this hash to remove duplicates...

Basically, we are trying to keep all that data around so one doesn't lose coverage on each deploy. In that sense, we need the data for each deployed version and we know to merge the data based on the MD5. Does that make sense?  We don't remove files as it would complicate things and instead recommend having a sane Redis TTL, something like 60 days or whatever fits your memory usage needs.

> Docker setup

I can try to share more later, I am happy to try to explain some of our setup, but it would take some time to put together anything super useful.

**Comment by danmayer** [comment]
hmm OK I might have finally been able to figure out / reproduce this in a test with the example data...

**Comment by danmayer** [comment]
excellent thanks for the data... basically it looks like since adding the MD5, first_updated_at, last_updated_at, and meta data feature merging data across multiple file paths as been broken. We hadn't seen any bug reports as it just doesn't seem to be as common as a usage pattern, but this functionally hasn't worked for the last 3 versions or so. I have a test replicating the problem now and should be able to put together a fix today.

**Comment by danmayer** [comment]
OK woohoo, figured it out... this was a pretty bad bug / regression, clearly, this feature didn't have enough test coverage and hasn't seen enough real-world usage for a while. I really appreciate all the support figuring this out @gingerlime 

take a look and we can make sure this is in the next release. We could put it in an alpha build release for you test and verify before wider release. 
https://github.com/danmayer/coverband/pull/191

**Comment by danmayer** [comment]
OK, @gingerlime if you want to verify the fix before release that would be awesome... I pushed the fix in a beta release. Should be as easy as bumping your gemfile to the below and `bundle install`

`gem 'coverband', '= 4.1.0.beta'`

let me know if you have any issues

**Comment by gingerlime** [comment]
@danmayer thanks a lot! that's awesome

I deployed `4.1.0.beta` on staging and it looks good again 👍 (unfortunately I won't have access to staging until the end of the week to test further across more deploys, etc)

> Basically, we are trying to keep all that data around so one doesn't lose coverage on each deploy. In that sense, we need the data for each deployed version and we know to merge the data based on the MD5. Does that make sense? We don't remove files as it would complicate things and instead recommend having a sane Redis TTL, something like 60 days or whatever fits your memory usage needs.

Wouldn't this blow up the data size by a pretty significant factor over time (especially if you deploy frequently)? I thought the data in redis is serialized into a single key, so if this is the case (is it?) then TTL won't work, because the redis data would stay fresh and just grow and grow... If data is split over many keys, then setting TTL could be a good solution. Where can I set this with coverband? I definitely want to keep coverage data across deploys, but not sure I'll get much value from coverage data for classes or modules that changed over time, or just copies of the same files across different folders.

btw, I checked and I still see all the keys for folders that were long removed from our server, so to me it doesn't make sense to keep those and could potentially fill up our redis + slow down coverband

```
pp Coverband.configuration.store.coverage.keys.select { |k| k.match(/dashboard_controller/) }
["/var/local/company/company.d/80/app/controllers/dashboard_controller.rb",
 "/var/local/company/company.d/79/app/controllers/dashboard_controller.rb",
...
 "/var/local/company/company.d/66/app/controllers/dashboard_controller.rb"]
```

Re Regex - understood. Not sure it's worth breaking backwards compatibility for.

Re Docker / ECS etc - would love to have a chat and hear more from you, if you have the time and energy, but I know it's asking for a lot.

**Comment by danmayer** [comment]
yes you are right the TTL model worked well previously when we had multiple keys per file... In the new storage for directories that change on deploy this would lead to unlimited growth...

I am pretty sure we can resolve that by merging the results to a single file when we merge coverage data, and throw out any data that is no longer mergable (IE md5 doesn't match). This definitely presents a bit of a unique challenge. I think some refactoring between merging results on end user reporting and merging during collection storage will be needed to keep this clean.

**Comment by gingerlime** [comment]
I guess until then, our best option is to clear the coverage data on every deploy, or every X deploys (although keeping track of it presents its own challenges). 

Is there anything I can help with? still not really familiar with the codebase, but with some pointers and help I might be able to figure out something.

**Comment by danmayer** [comment]
yeah I think until then clearing every X days is likely the best. Sorry about this @gingerlime definitely another thing missed when we changed our storage mechanism. I don't think it should be too bad to add the merge in... basically our save mechanism works like so

* pull old coverage report from
* merge with new data
* store new coverage report on redis

The "merge with new data" step currently merges assuming this only needs to handle match filenames, because this was intended as a runtime merge, just updating code hits in the last 30 seconds... I think that step could also handle merging file changes that match `roots` patterns. If we do this every time it would never get more than 2 levels deep (old and new report about to be merged)...

The only real concern is this moves it from a report time perf cost to a runtime perf cost. We mostly deal with that by it being in a background thread, but that is still a concern. The other idea I would have would be having a rake job that could be run periodically to 'flatten' the store.

thoughts @gingerlime

**Comment by gingerlime** [comment]
@danmayer interesting. I guess it's happening [here](https://github.com/danmayer/coverband/blob/2f0785eac742c0d28d2e55761f78eb573f309712/lib/coverband/adapters/base.rb#L64).

Canonicalizing the paths could be a solution, but could incur some performance penalty. I wonder however, if merging can be based on md5 hash instead of the filename key? It would change the merge algorithm, but I think performance is likely to be similar (not sure, but intuitively it's still same order of magnitude). I think you already have some benchmarks in the codebase, so those can help gauge if it's the same ballpark. One thing to consider is that as the data grows now, every merge takes longer... So by merging duplicate entries, we're reducing performance over time. But of course, this only applies to deployments where the base path changes.

Doing a separate canonicalization/flattening/squashing via a rake task that can run periodically is also possible, I guess. I think there's a good chance that the performance hit doing it "inline" is marginal, and it's more elegant this way. But it's just a guess at this point.

Is there some kind of a pre-made development/test docker image to play around with?

**Comment by danmayer** [comment]
> I wonder however, if merging can be based on md5 hash instead of the filename key? It would change the merge algorithm, but I think performance is likely to be similar (not sure, but intuitively it's still same order of magnitude)

Yeah this would be possible, shouldn't be to much slower, just have to dig into the hash value vs merging entirely based on key. 

> One thing to consider is that as the data grows now, every merge takes longer... So by merging duplicate entries, we're reducing performance over time. But of course, this only applies to deployments where the base path changes.

If we merged every time it should always flatten from 2 -> 1 depth right? So I think it would just be a constant hit vs one that grows over time.

> Is there some kind of a pre-made development/test docker image to play around with?

No generally I use coverband_demo (https://github.com/danmayer/coverband_demo) with it's gemfile pointed at my local machines Coverband. For testing that works pretty well... I could try to set up a quick and easy Docker to make it even easier for folks to contribute.

I guess this issue, is probably ready to be closed but we want to open a new one for the constant grown when deploying to changing directories?

**Comment by gingerlime** [comment]
Yeah, I think it should be a separate issue as well.

Perhaps a silly question, but where would the original folder paths come from? is it something that coverage generates for coverband? I wonder if it's also possible/preferable to canonicalize the folders at that point? rather than during merge?

In any case, probably best to discuss options on the new issue :) I'll open a new one and we can continue there?

**Comment by gingerlime** [comment]
created #193. Will close this for now :)

Thanks again @kbaum and @danmayer for your incredible support.
